### DISCUSSION TITLE
Simpleperf on Pixel 6 Android 13

### DISCUSSION THREAD

**Discussion by hosunnyc** [android/ndk#1833] [Q&A] [answered]
Hi, I just wanted to ask about `simpleperf list` on a rooted Pixel 6 Android 13 OS.
Honestly, this might not be the right place to ask this but I couldn't find another area to ask a device specific simpleperf tool, so if there is a better place, I can try there.

There is a lot of `pmu events` listed under `simpleperf list` like `aoc/cycles/`, `tpu/cycles/`, `gpu0/cycles/`, and thousands more. However, when I do try to retrieve those counters at 0.1s intervals, my Pixel 6 would either crash or the values would continuously be 0 even though I feel positive that the module is being used.

Does this mean that these events are actually not supported? Or should I ask this in a different forum with the developers behind Pixel 6?
Thanks!

**Comment by stephenhines** [comment]
@yabinc might be able to answer some of these questions.

**Comment by yabinc** [comment, marked as answer]
This is the right place to ask simpleperf questions, as listed in https://android.googlesource.com/platform/system/extras/+/master/simpleperf/doc/README.md#bugs-and-contribution :)

The PMU events listed by simpleperf is collected by reading /sys/bus/event_source/devices, which is currently only readable when we are root. It lists all PMUs registered via perf_pmu_register() in the kernel, as in https://elixir.bootlin.com/linux/latest/ident/perf_pmu_register.

Some PMUs are common to all arm64 devices, like armv8_pmuv3 (the CPU PMU), or cs_etm (for ETM data collection). But some
PMUs are device specific. They may have special way to use, or may crash when used not properly, or not work at all. And it is likely that these PMUs don't have publicly available kernel code or documentation. The device specific PMUs are usually only used for system wide performance measuring internally. That's why we need to be root to access them. So I don't suggest using them unless you know what you are doing.

**Reply by hosunnyc** [reply]
So I guess even with root access, just because they're listed on Simpleperf does not necessarily mean that we can use them since they have specified ways of usage, which is for the internal Pixel 6 developers?
As for the common PMUs on all arm64 devices, I'm guessing I would have to figure that list out on my own then?

**Reply by yabinc** [reply]
Right. There is no test to guarantee that each device specific pmu event works.
For common PMUs, one way to find them is to search  perf_pmu_register() in the common kernel. If you can find a PMU in the common kernel, it should be the same to all arm64 devices. Otherwise, the PMU probably only appears in a device specific kernel.

**Reply by hosunnyc** [reply]
Hmmmm I see alright. Thanks.
Also, is there a way for Simpleperf to list the clock cycle at which the counter is sampled instead of 0.1023s, 0.2014s, etc in the `Total test time` area when using `stat` at 0.1s intervals? Or maybe another method?

**Reply by yabinc** [reply]
The 0.1023s, 0.2014s are monotonic clock time elapsed since start of the stat command.
For clock cycle, do you mean to print timestamp in monotonic clock?

**Reply by hosunnyc** [reply]
Right yeah, so instead of the monotonic clock time elapsed since the start. Just printing the raw monotonic clock value.

**Reply by hosunnyc** [reply]
Because on simpleperf record, if you set `clockid` as monotonic, the timestamps in `report` sample are shown as monotonic clock instead of 0.1023s, etc. If thats not supported under `stat`, then can `record` do the same thing as `stat` where raw PMU events are sampled with options like `--per-cpre` and `--interval-only-values`?

**Reply by yabinc** [reply]
> instead of the monotonic clock time elapsed since the start. Just printing the raw monotonic clock value.

It's possible. But I want to know the reason you want the change. I also suggest to create a separate issue.
The timestamp is generated in userspace using std::chrono::steady_clock::now(). In case you want to
make the change by yourself, the code is in https://android.googlesource.com/platform/system/extras/+/master/simpleperf/cmd_stat.cpp#587.

>  can record do the same thing as stat where raw PMU events are sampled with options like --per-core and --interval-only-values

This isn't needed. Because in a sample, the event value (period) is the number of events happened since the previous sample on the same cpu. So it already has the effect of --per-core and --interval-only-values.

**Reply by hosunnyc** [reply]
Ahhh alright. I mostly needed it to synchronize with another application, but it seems simpleperf record can do something similar so might not be necessary! Thanks.
